pkg/cloud_provider/user_tags.go

```diff
@@ -0,0 +1,523 @@
+		return fmt.Errorf("failed to fetch %s tag: %w", name, err)
+	}
+	return nil
+}
+
+// close the tag binding client connection to API server.
+func (c *tagBindingsClient) close() {
+	if err := c.Close(); err != nil {
+		klog.V(1).ErrorS(err, "failed to close tag binding client connection")
+	}
+}
+
+// getEffectiveTagList returns the list of tags attached to the resource,
+// which were inherited or attached to the resource directly.
+func (c *tagBindingsClient) getEffectiveTagList(ctx context.Context, rscName string) resourceTags {
+	effectiveTags := make(resourceTags)
+	bindings := c.listEffectiveTags(ctx, rscName)
+	// a resource can have a maximum of {gcpMaxTagsPerResource} tags attached to it.
+	// Will iterate for {gcpMaxTagsPerResource} times in the worst case scenario, if
+	// none of the break conditions are met. Should the {gcpMaxTagsPerResource} be
+	// increased in the future, it should not create an issue, since this is an optimization
+	// attempt to reduce the number of tag write calls by skipping already existing tags,
+	// since tag write operation has quota restriction.
+	for i := 0; i < maxTagsPerResource; i++ {
+		binding, err := bindings.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil || binding == nil {
+			// on encountering any error will continue adding refined tags
+			// which will still have all the user provided tags except for
+			// the removed already existing tags processed until this point
+			// and would end up adding tags which already exist, and error
+			// handling is present for the same.
+			klog.V(5).Infof("failed to list effective tags on the %s resource: %v: %v", rscName, binding, err)
+			break
+		}
+		effectiveTags[binding.NamespacedTagValue] = resourceTagsValueFiller
+	}
+	return effectiveTags
+}
+
+func (c *tagBindingsClient) listEffectiveTags(ctx context.Context, name string) *rscmgr.EffectiveTagIterator {
+	return c.ListEffectiveTags(ctx, &rscmgrpb.ListEffectiveTagsRequest{
+		Parent: name,
+	})
+}
+
+// getTagsToBind returns the filtered list of tags after removing the tags which
+// already exist on the resource.
+func (c *tagBindingsClient) getTagsToBind(ctx context.Context, rscName string, tags resourceTags) (resourceTags, error) {
+	effectiveTags := c.getEffectiveTagList(ctx, rscName)
+	refinedTags := make(resourceTags)
+
+	for tag := range tags {
+		if _, exist := effectiveTags[tag]; !exist {
+			refinedTags[tag] = resourceTagsValueFiller
+		}
+	}
+	klog.V(1).Infof("refined list of resource tags: %v", refinedTags)
+
+	return refinedTags, nil
+}
+
+// createTagBindings creates the tag bindings for the resource.
+func (c *tagBindingsClient) createTagBindings(ctx context.Context, rscName string, tags resourceTags) error {
+	if len(tags) == 0 {
+		return nil
+	}
+
+	// GCP has a rate limit of 600 requests per minute, restricting
+	// here to 8 requests per second.
+	limiter := newRequestLimiter(tagsRequestRateLimit, tagsRequestTokenBucketSize, true)
```

Comment:
Does this need to be that restrictive to 8 rps? I think you increase this beyond that, and use a rate limiter with exponential backoff/retry to mitigate the quota delay.

Also, what happens if we get an API error due to quota exhausted (eg: how would we retry to re-add these tags to the resource)? I think we'd want to requeue and retry, right now it seems like if we hit the 600 per-minute-limit, all subsequent requests in this loop would likely fail.